packages/proxy/src/handler.ts

```diff
@@ -66,40 +63,73 @@ function isRedirect(
 
 export async function handler(event: CloudFrontRequestEvent) {
   const { request } = event.Records[0].cf;
+  const domainName = request.origin!.s3!.customHeaders['x-env-domain-name'][0].value;
   const configEndpoint = request.origin!.s3!.customHeaders[
     'x-env-config-endpoint'
   ][0].value;
-  const apiEndpoint = request.origin!.s3!.customHeaders['x-env-api-endpoint'][0]
+  let apiEndpoint = request.origin!.s3!.customHeaders['x-env-api-endpoint'][0]
     .value;
```

Comment:
The `!`s here could be made more safe with validation.

If for whatever reason these things do not exist, a contextual error message is going to be a lot more helpful in an operation context than e.g. `"cannot read 'x-env-domain-name' of undefined"`.

Since this code is deep infra code, I would say going the extra mile is worth it here.

Other option here is give a good reason why `!` is acceptable in these cases via a code comment.

An ESLint rule to catch use of `!` might be helpful.

Then you can disable the ESLint check on a case-by-case basis with an ignore comment + comment about why, which is a nice variation IMO of the above suggestion to add a comment generally if the `!` is actually correct.

Reply:
This was based on the original code, but I now changed it so it checks that the chain actually exists and also added a comment explaining why we expect it to be there.